### ISSUE TITLE
Implement FluidSynth integration

### ISSUE THREAD

**Issue by dreamer** [dosbox-staging/dosbox-staging#262]
The patch integrating FluidSynth was developed a long time ago (first revision 9 years ago, last revision 3 years ago - it really waits that long for inclusion upstream?). It is currently being distributed via ECE.

In the past, I was very hesitant to integrate FluidSynth patch for a number of reasons:

1. I wanted to clean up first, before starting new work in the area. Without going into details… *there are serious issues with DOSBox MIDI support* - the demand for integrating Fluid might be caused by those issues.
2. I don't want to target Fluid 1.x API. It's old - Fluid is being actively developed and they have 2.x API released already.
3. When developing Boxtron I tested FluidSynth running as application thoroughly… and the results were very mixed - to the point, that I recommend Boxtron users to use Timidity++ by default. All that testing was around version 1.1 though (which is old and hopefully not representative of quality in 2.x).
4. After a brief look at patch itself, I have some reservations: MIDI synth should not run in the same process as emulation loop, properties do not use the same naming style as other options, INSTALL documentation is missing, user documentation is missing, build was never tested under Visual Studio, (and more) - all those issues can be easily fixed; patch itself is not that big nor invasive, *it just needs some work*.
5. libfluid 2.x was not available in repositories, which would mean integrating patch required bundling fuid in the repo, which means transition to cmake, yada, yada - all the same reasons as for #257 

### What changed?

Over last 1-2 months (?), point (5) stopped being a problem, [fluidsynth 2.x is now available in most repositories, including brew and vcpkg](https://repology.org/project/fluidsynth/versions). That takes out reason (5), and it's huge.

Points (4) and (2) are just a matter of putting in some work - not a big deal. I still want to address (1), but after looking in details at the patch - this work can largely happen in parallel. (3) can be assessed only during testing.

Of course, integration of Fluid needs to be optional - the lib is not old nor tested enough to be propagated to all repositories, so some users will want to build without it. But the patch already kind-of handles it.

### When?

Testing effort required would push 0.75.0 release too far away, so I won't merge any new MIDI-related changes before 0.75.0 release ([here are tasks planned ATM for 0.75.0](https://github.com/dreamer/dosbox-staging/projects/2)).

But if someone has free cycles and wants to push this work forward: the patch is imported on a side branch already - cherry-pick f786b9fdc52df0e9b39ee27d640fb37ba7c369e6 and start working on fixes to problems I listed in (2) and (4) ;) If nobody will step up, I'll start work on this sometime after 0.75.0.

**Comment by kcgen** [comment]
Well laid-out plan @dreamer ! :bookmark_tabs:

**Comment by bluddy** [comment]
Regarding part of point 4, I assume you mean using separate threads for midi synthesis. Fluidsynth already uses other threads internally, meaning that it doesn't burden the dosbox thread.

In general, the patch seems simple enough to me that it can just be merged. Further cleanup can proceed later on as a specific task. Most importantly, it'll allow more users to switch to dosbox-staging, who will then be invested in the project's success.

**Comment by dreamer** [comment]
> Fluidsynth already uses other threads internally, meaning that it doesn't burden the dosbox thread.

We need proof to confirm this assertion. It's not hard, just generating [flamegraphs](http://www.brendangregg.com/flamegraphs.html) and some testing - but that's part of development work needed to push the patch through the finish line.

Also, I am not going to merge patch in a state where CI does not verify the build using multiple compilers on multiple OSes. And that's the state the patch is in ATM :(.

**Comment by bluddy** [comment]
> We need proof to confirm this assertion. It's not hard, just generating flamegraphs and some testing - but that's part of development work needed to push the patch through the finish line.

From the [fluidsynth pages](http://www.fluidsynth.org/api/):
> As soon as the audio driver is created, it will start playing. The audio driver creates a separate thread that uses the synthesizer object to generate the audio.

This is unlike munt, which has no notion of threading.

**Comment by kcgen** [comment]
> Most importantly, it'll allow more users to switch to dosbox-staging

Great point @bluddy , I agree.  Regarding the flamegraphs @dreamer mentioned, are you able to generate them?  

I haven't had time to dig in yet; but if you can build and run fluidsynth (standalone) on a linux box, here are notes on how to generate them: https://github.com/brendangregg/FlameGraph/blob/master/README.md

**Comment by dreamer** [comment]
Without going into details, this is the way I generate flamegraphs (Linux only), before starting the FlameGraph repo mentioned earlier needs to be cloned (for the scripts inside).

1. Build dosbox-staging with `-g -O0 -fno-omit-frame-pointer`
2. Start dosbox-staging, and note PID of the process
3. Start the benchmark / test inside dosbox, and immediately:
4. `sudo perf record -F 99 -p <PID> -g -- sleep 120` (-F 99 is probing stack at 99Hz - empirically, this seems like a good middle-ground, at least for me (we don't want too high value for probing, as it makes the results less realistic!); play the game or watch benchmark for 2 minutes
5. `sudo perf script | ~/src/FlameGraph/stackcollapse-perf.pl > out.perf-folded`
6. `~/src/FlameGraph/flamegraph.pl out.perf-folded > flamegraph-game-test-description.svg`

Resulting SVG to be opened in the browser (it has js to allow for easier browsing and filtering the graph). I need to describe somewhere how to interpret the graph, but overall: there will definitely be a huge plateau of unrecognized stacks (that's dynrec-generated CPU emulation, we're not interested in that) and a tower, that can be clearly recognized as main dosbox "Normal" loop; narrow towers on top of normal loop are ok - plateaus on top of normal loop are bottlenecks. (The thing I'm not sure about is how to be 100% sure flamegraph does not include child threads - but for CD-DA this doesn't seem to be a problem).

If flamegraphs with FluidSynth integrated will look similar to graphs when game is playing music via CD-DA emulation - that's good. if fluidsynth will show up as plateaus covering ~5% of runtime stacks or more - that's bad and we'll need to investigate further.

**Comment by kcgen** [comment]
Instructions are great! Here's a zoom out of what I got (Jones in the Fast Lane, ~2500 cycles, FLAC CD-DA sequences back-to-back)

![Screenshot at 2020-04-14 15-03-52](https://user-images.githubusercontent.com/1557255/79278725-2f607700-7e61-11ea-9eb4-6c88a9300e82.png)

**Comment by dreamer** [comment]
Wow, this looks completely different than what I get on x86_64 - can you paste plain svg somewhere?

**Comment by dreamer** [comment]
Work on this feature started on branch `po/fluid-1`. Old version of this patch as distributed via ECE has some issues - we are going to use @realnc implementation instead ([link](https://github.com/realnc/dosbox-core/blob/libretro/libretro/midi_fluidsynth.cpp)) - it seems cleaner, was already converted to FluidSynth 2.1, and seems to be better tested. But we will need to do some tweaks anyway, as certain small design choices in there clash with our future plans.

**Comment by kcgen** [comment]
A recent platform-specific timer adjustment for fluidsynth: https://github.com/joncampbell123/dosbox-x/commit/e00cf22392f9c0eb9bef3e3e8edea4fa433dc609

https://www.vogons.org/viewtopic.php?t=27831&start=60

**Comment by realnc** [comment]
> A recent platform-specific timer adjustment for fluidsynth: [joncampbell123/dosbox-x@e00cf22](https://github.com/joncampbell123/dosbox-x/commit/e00cf22392f9c0eb9bef3e3e8edea4fa433dc609)
> 
> https://www.vogons.org/viewtopic.php?p=852446#p852446

These are only used when fluidsynth is doing audio output itself. They are audio driver parameters. When rendering audio into a buffer (with `fluid_synth_write_s16()` in this case) and letting the dosbox mixer play the audio without creating a fluidsynth audio driver, these parameters have no effect whatsoever.

http://www.fluidsynth.org/api/index.html#UsingSynth

This is how it's done in dosbox-core. The fluidsynth patch that's been floating around for a while now for vanilla dosbox does not do this, and thus there these parameters are important.

**Comment by kcgen** [comment]
Thanks for the comparison @realnc, and good to know these adjustments won't be needed.

The approach of feeding dosbox's mixer with samples is win-win-win: fewer LOC to maintain, uses a single host-agnostic audio interface abstracted by SDL, and less runtime complexity and overhead.

**Comment by dreamer** [comment]
Initial, working version of FluidSynth integration can be tested on branch `po/fluid-3` - at this point it is (almost) direct port from dosbox-core, but using our normal coding conventions, licensing info and SPDX identifier. Code was also moved to the recently created `midi` module, to avoid littering `gui` any more.

**Testers**: *you need to compile it yourself*. Our CI does not provide precompiled snapshots with FluidSynth integration (yet). FluidSynth 2.x is available in many distro repositories, but it's still missing from a few notable ones.

Do not ask me for support as of yet - you're on your own. Do not get married to new fluid settings as inherited from dosbox-core - we will change them (not sure exactly how yet).

I tested it on Ubuntu 20.04 and Windows 10 and it seemed to work *fine*, but code is not good enough quality-wise and we have no user documentation, so it won't be merged to master just yet.

**Comment by dreamer** [comment]
The first part of FluidSynth 2.x support was just merged via #539 :)

But I'm not closing this feature request just yet - we need to polish it a little bit, add more documentation, implement some missing bits, we have 1 small bug… but as of now, dosbox-staging finally has a built-in MIDI synth.

To testers: recreate your config file - there's a new `fluidsynth` section in. The current set of user-changeable settings is not final.

I am especially interested in learning from testers using wide range of SoundFonts:

- Is MIDI too loud or too quiet? Does it change with SoundFonts tested?
- Is it too CPU intensive? Is playback smooth?
- Is memory consumption appropriate? We're considering tweaking FluidSynth to load samples dynamically to lower the memory consumption (or maybe allowing users to change it)
- Are you missing something specific?
- What SoundFonts do you like and use?

If you're compiling the code yourself, then FluidSynth support should work on any OS.
If you're using our pre-compiled snapshot builds, then ATM only Windows builds have the feature enabled (fluidsynth 2.x library is missing from brew repo on macOS and from Ubuntu 18.04 repos, so we cannot provide pre-compiled packages on those OSes yet).

**Comment by arrowgent** [comment]
i think this is from SVN
source dosbox: https://launchpad.net/~i30817/+archive/ubuntu/dosbox-patched

```
##
#            fluid.driver: Driver to use with Fluidsynth, not needed under Windows. Available drivers depend on what Fluidsynth was compiled with
#                            Possible values: pulseaudio, alsa, oss, coreaudio, dsound, portaudio, sndman, jack, file, default.
#         fluid.soundfont: Soundfont to use with Fluidsynth. One must be specified.
#        fluid.samplerate: Sample rate to use with Fluidsynth.
#              fluid.gain: Fluidsynth gain.
#         fluid.polyphony: Fluidsynth polyphony.
#             fluid.cores: Fluidsynth CPU cores to use, default.
#           fluid.periods: Fluidsynth periods.
#        fluid.periodsize: Fluidsynth period size.
#            fluid.reverb: Fluidsynth use reverb.
#                            Possible values: no, yes.
#            fluid.chorus: Fluidsynth use chorus.
#                            Possible values: no, yes.
#   fluid.reverb,roomsize: Fluidsynth reverb room size.
#    fluid.reverb.damping: Fluidsynth reverb damping.
#      fluid.reverb.width: Fluidsynth reverb width. (.76)
#      fluid.reverb.level: Fluidsynth reverb level. (.57)
#     fluid.chorus.number: Fluidsynth chorus voices
#      fluid.chorus.level: Fluidsynth chorus level.
#      fluid.chorus.speed: Fluidsynth chorus speed.
#      fluid.chorus.depth: Fluidsynth chorus depth.
#       fluid.chorus.type: Fluidsynth chorus type. 0 is sine wave, 1 is triangle wave.
#                            Possible values: 0, 1.
```


i use a Roland SC-55 soundfont, i cant remember where i got it from
FluidR3_GM_sc-55.sf2 108424522 bytes

i dont have fluidsynth running all the time
i call it to start using a script then whatever application requires it
when the application exits so does fluidsynth.
ie:
```
/usr/bin/fluidsynth -a pulseaudio -m alsa_seq -i -l -s -p FluidSynth /usr/share/sounds/sf2/FluidR3_GM.sf2 &
/usr/games/dosbox &&
```

that said, i dont start fluidsynth with dosbox-staging
ive only done minor testing with dosbox-staging, sounds seem to be fine
there were a few games which music was not playing.  i will try to figure out which games do/dont play
both by starting fluidsynth and by using static built-in fluidsynth

**Comment by arrowgent** [comment]
some dos midi software useful for testing?
some midi players could be useful for exclusively testing midi (no extra game/sfx/etc, and lightweight)

http://dosmid.sourceforge.net/

also GSPLAY 1.0
version gsplay1.zip or labelled 1.1
i dont see any public sources, although many respositories list it as "freeware" so i wont post a link here
not the same as gsplay v2.x free which is for windows

untested but others report using:
megamid dos

**Comment by kcgen** [comment]
With @grapeli's report of playback gaps on slightly older hardware, I wondered if the integrated FS library really is as threaded as an external midi player is when processing sysex calls entirely out of band relative to dosbox.

The documentation http://www.fluidsynth.org/api/ mentions:

> FluidSynth's rendering engine is implemented by using the "Dispatcher Thread Pattern". This means that a certain thread A, which calls one of FluidSynth's rendering functions, namely
> 
> fluid_synth_process()
> fluid_synth_nwrite_float()
> fluid_synth_write_float()
> fluid_synth_write_s16()
> automatically becomes the "synthesis thread". The terms "synthesis context" and "synthesis thread" are equivalent

So.. we have Dosbox's 1ms loop calling FS's above-mentioned rendering function, and this means the synthesis is actually be done in Dosbox's main thread.

That was true for 2.1.5, which is the latest.  The docs mentions for 2.2.0:

> The sequencer has received a major revisal. For you that means:
> The sequencer's queue no longer blocks the synthesizer thread, due to being busy arranging its events internally.

So I guess some underlying work, namely this "sequencer queue that arranges events" is now being broken out of the rendering call.. so hopefully that further reduces the block-time when synthesizing the audio. 

I suspect to fully disconnect FS from blocking us, we'd need to put the FS synth object inside a thread, and front-run it's rendering by some number of milliseconds, so when Dosbox's 1ms loop comes around the samples are ready to be written into the audio channel.

**Comment by dreamer** [comment]
So just as we suspected in the beginning… :( Also, notably, 2.2.0 is ABI incompatible with previous versions… we will need to keep an eye on this.

I'm afraid dosbox-staging mixer is not prepared to handle callbacks from a different thread if we'll move FS sequencer to a separate thread. But I haven't looked in mixer code in detail yet, I hope I'm wrong.

**Comment by grapeli** [comment]
```
[fluidsynth]
synth_threads = 1

# ps -T -l -C dosbox
F S   UID     PID    SPID    PPID  C PRI  NI ADDR SZ WCHAN  TTY          TIME CMD
0 R  1000   31240   31240    1368 24  89   0 - 195520 -     pts/6    00:00:04 dosbox
1 S  1000   31240   31241    1368  0 107  19 - 195520 -     pts/6    00:00:00 dosbox:disk$0
1 S  1000   31240   31242    1368  0 107  19 - 195520 -     pts/6    00:00:00 dosbox:disk$1
1 S  1000   31240   31243    1368  0 106  19 - 195520 -     pts/6    00:00:00 dosbox:disk$2
1 S  1000   31240   31244    1368  0 107  19 - 195520 -     pts/6    00:00:00 dosbox:disk$3
1 S  1000   31240   31245    1368  0  99  19 - 195520 -     pts/6    00:00:00 SDLHotplugALSA
1 S  1000   31240   31246    1368  1  80   0 - 195520 -     pts/6    00:00:00 SDLAudioP2

[fluidsynth]
synth_threads = 2

# ps -T -l -C dosbox
F S   UID     PID    SPID    PPID  C PRI  NI ADDR SZ WCHAN  TTY          TIME CMD
0 S  1000   31194   31194    1368 45  80   0 - 214010 -     pts/6    00:01:01 dosbox
1 S  1000   31194   31195    1368  0 106  19 - 214010 -     pts/6    00:00:00 dosbox:disk$0
1 S  1000   31194   31196    1368  0 106  19 - 214010 -     pts/6    00:00:00 dosbox:disk$1
1 S  1000   31194   31197    1368  0 106  19 - 214010 -     pts/6    00:00:00 dosbox:disk$2
1 S  1000   31194   31198    1368  0 106  19 - 214010 -     pts/6    00:00:00 dosbox:disk$3
1 S  1000   31194   31199    1368  0  99  19 - 214010 -     pts/6    00:00:00 SDLHotplugALSA
1 S  1000   31194   31200    1368  1  80   0 - 214010 -     pts/6    00:00:01 SDLAudioP2
1 S  1000   31194   31201    1368  4  -1   - - 214010 -     pts/6    00:00:06 mixer0
```

In this case, the priority of the mixer thread in htop is `-61` (very high, the same is also external fluidsynth).

**Comment by kcgen** [comment]
@grapeli , thanks :👍

Assuming we will have other users in this boat, please share your recommended config changes that work better than our current defaults (no rush; once you're done investigating). 

If these also work for others without detriment, then I suggest we go with your settings as safe/better values while we address the remaining issues with FluidSynth (such as putting it in parallel thread instead of blocking our main loop).

**Comment by kcgen** [comment]
Below comment from @grapeli moved from PR https://github.com/dosbox-staging/dosbox-staging/pull/640#issuecomment-706623816 to this FluidSynth discussion thread.

> My problems with the built-in fluidsynth are too slow access to data on the disk (hdd). Which can cause the sound hiccups generated by the built-in fluidsynth. Changes to the mixer settings do not affect this (this is my completely misinterpretation). The real reason for smooth audio is that the second, third, fifth time the data is cached or completely in memory (tmpfs).

> CPU is not a problem, although in HoMM2 there may be a load spike close to this limit. I built an optimized (profiled) version of this branch, with interpolation, chorus and reverb turned off (the sound is clearly lower quality), but the load is also lower.

**Comment by kcgen** [comment]
Interesting @grapeli .  So when you move your SF2 to /dev/shm (or tmpfs), the problem disappears?

In this case @dreamer, we should consider a follow on PR that reads the entire SF2 into memory. Something like: http://www.fluidsynth.org/api/fluidsynth_sfload_mem_8c-example.html

However, this solution is pretty ugly. Ideally FluidSynth would offer an option to read the entire file into memory on load to eliminate poor IO latency from DoS'ing the stream.  I will open and issue and see what their developers think.

> CPU is not a problem, although in HoMM2 there may be a load spike close to this limit.

Yes - the two are often connected (plus Linux reports blocked  IO as CPU user time to :sweat_smile:).
This was also happening to be on the Pi (I think we had a discussion about this); and on there I had to tell Timidity to load the entire SF2 into RAM, and I inculded a delay before launching DOSBox to account for this pre-load duration.  With that technique, I would lose several seconds of MIDI durnig the Sierra logo startup sequence.

We definitely need to find the equivalent solution with FluidSynth.

 > I built an optimized (profiled) version of this branch, with interpolation, chorus and reverb turned off (the sound is clearly lower quality), but the load is also lower.
 
Thanks for these tests.  Just to be absolutely clear, can you confirm that you have zero-hiccups when the SF2 is loaded from /dev/shm/path/ (RAM-disk), even with chorus, reverb, and 7th order poly all enabled?  If you confirm this, then that would confirm that CPU-load-wise we're are good and that IO is exclusively the problem (even on older hardware).

**Comment by kcgen** [comment]
@grapeli , see: https://github.com/FluidSynth/fluidsynth/issues/685

**Comment by grapeli** [comment]
> Interesting @grapeli . So when you move your SF2 to /dev/shm (or tmpfs), the problem disappears?
> 

I ran another test.
1. Soundfont on hdd disk and one directory with DOS programs including HoMM2.
[01-homm2.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5360830/01-homm2.flac.txt)
Approximately 26 audio interruptions.

2. Soundfont on hdd, HoMM2 mounted as a separate directory from hdd.
[02-homm2.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5360834/02-homm2.flac.txt)
Three breaks, only in the first 10 seconds.

3. Soundfont on hdd, HoMM2 mounted as a separate directory from tmpfs.
[03-homm2.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5360835/03-homm2.flac.txt)
Zero interruptions in audio delivery.

Each test is preceded by cleaning the cache.
`echo 3 | sudo tee /proc/sys/vm/drop_caches`

**Comment by dreamer** [comment]
OK, so as for now we know:

1. FluidSynth loads whole soundfont into the memory, processes it and closes the handle. It does that because we keep `synth.dynamic-sample-loading` to default (disabled). We discussed enabling this feature previously, but this behaviour indicates we should probably keep it as disabled. Or maybe even explicitly set it to false to indicate it's disabled by design.
2. @grapeli findings are very interesting in this context and warrant further investigation, but first we'll need to address (3)
3. FS maintainers confirmed we need to create separate synthesis thread - this is the most likely source of issues right now.
4. WIP: Soft Limiter task to mitigate impact of too loud/too quiet soundfonts.
5. WIP: Loading SF files from known locations - some prerequisites are being implemented.

@kcgen I'll wait with starting work on (3) until SoftLimiter work will be finished to avoid conflicts.

**Comment by mawe42** [comment]
> 1. FluidSynth loads whole soundfont into the memory, processes it and closes the handle. It does that because we keep `synth.dynamic-sample-loading` to default (disabled). We discussed enabling this feature previously, but this behaviour indicates we should probably keep it as disabled. Or maybe even explicitly set it to false to indicate it's disabled by design.

Just one more quick comment regarding the dynamic-sample-loading: it might be worthwhile to give your users the option to choose dynamic loading vs. static loading of the samples. Especially when using many or large Soundfonts, enabling it will help to save possibly hundreds of megabytes of RAM in addition to reducing the load time of the Soundfonts significantly. The big downside is IO in the render thread, of course. But my guess is that you are using FS only for music playback, so you could probably get away with a large buffer and latency in the 100+ ms range. And with that, dynamic sample loading will probably only be problematic on really old hardware or very slow HDDs...

Anyway, I hope you get FS integrated into dosbox without problems. In case you need more input, the [fluid-dev mailing-list](https://lists.nongnu.org/mailman/listinfo/fluid-dev) is the best place for general feedback and usage questions.

**Comment by kcgen** [comment]
@grapeli, 

Your tests match with what the FluidSynth team explaned (that FluidSynth by default reads the entire SF2 into memory),  which explains why there is no stuttering in tests 3 and 4.

I was able to measure this full-load behavior using the ~517 MiB SF2:
 - **Left pane**: `sudo pmap <dosbox-staging PID>` shows a single heap allocation holding the SF2
 - **Right pane**: `dstat` shows the disk reads per second accumating to ~519 MiB, after which the dosbox shell appears

![2020-10-11_09-32](https://user-images.githubusercontent.com/1557255/95684557-6234bf00-0ba7-11eb-8327-1ade55483cef.png)

I agree with @dreamer's 2nd point above because an IO stall inside dosbox's emulation loop (in this case, reading a HOMM2 data file as shown in your tests 1 and 2), will block dosbox from gathering audio from its channel sources; because dosbox employs one big serialized loop.  

@dreamer - even if we put all of FluidSynth into its own dosbox thread, dosbox's channel callback will still be operating in dosbox's main loop (this is the callback that copies the sample buffer into the mixer via `add_samples_(...)`).  Likewise, once all the callbacks have fed dosbox's mixer, it combines them into a final mixed audio buffer and writes that out to SDL..  which will also be blocked because the mixer is part of the serial loop.  Ugh.. so the mixer also needs to be made asynchronous too if this is going to work.

**Comment by MasterO2** [comment]
> 
> 
> @grapeli,
> 
> Your tests match with what the FluidSynth team explaned (that FluidSynth by default reads the entire SF2 into memory), which explains why there is no stuttering in your tests 3 and 4.
> 
> I was able to measure this load-in-full behavior using the ~517 MiB SGM soundfont:
> 
>     * **Left pane**: `sudo pmap <dosbox-staging PID>` shows a single heap allocation holding the soundfont
> 
>     * **Right pane**: `dstat` shows the disk reads per second accumating to ~519 MiB, after which the dosbox shell appears
> 
> 
> ![2020-10-11_09-32](https://user-images.githubusercontent.com/1557255/95684557-6234bf00-0ba7-11eb-8327-1ade55483cef.png)
> 
> @dreamer, I agree with your 2nd point above because an IO stall inside dosbox's emulation loop (in this case, reading a HOMM2 data file as shown in @grapeli's tests 1 and 2), will block dosbox from gathering audio from its channel sources; because dosbox employs one big serialized loop.
> 
> To expand on it: even if we thread FS's synth call, dosbox's channel callback will still be operating in dosbox's main loop (that is, the callback that copies the buffer produced by `synth(..)` into the mixer via `add_samples_(...)`). Likewise, once all the callbacks have fed dosbox's mixer, it combines them into a final mixed audio buffer and writes that out to SDL, which will also be blocked because the mixer is part of the serial loop.
> 
> So the mixer, its callbacks (and something to drive them on a timer), and FluidSynth all need to be threaded if we want SDL to keep receiving audio samples (and thus keep hearing MIDI music) while dosbox's primary loop gets blocked on game IO stalls.
> 
> This is going to be large architectural overhaul where all of the emulated sound devices would become threaded. I'm not sure how it will work, given many of those sound devices require the DOS program to be poking and prodding their registered ports and IRQs, which influence generated audio.

@kcgen So that architectural overhaul is for after 0.76?

**Comment by kcgen** [comment]
@MasterO2 - I'm not sure.

Moving FluidSynth to its own thread could happen in time for 0.76 (@dreamer's #2 - and nothing more). But I don't think that will solve the issue.

I'm very worried that threading dosbox's entire audio subsystem will result in broken audio for devices that involves lock-step port & IRQ control (ie: Sound Blaster, Adlib, and GUS); so we'd need a hybrid approach where only the relatively stand-alone audio sources (such as MIDI and CDDA) are broken out with a separate threaded path out SDL; and I personally don't have the stomach for that amount of work (at least at this point), so I can't vouch for a timeframe.

That said, there are other zero-risk approaches that might help @grapeli :-) Working on that right now.

**Comment by grapeli** [comment]
@kcgen 
I more or less understand the cause of the problem.

I will add that I did one more test (four times) with the external fluidsynth for the most demanding first case. The result for four repetitions is 100% correct.

I checked what files are read during the test.
`inotifywait -m -e access -e open -r HoMM2/ 2>&1 | tee /tmp/hero2.log`
Mostly this one DATA/HEROES2.AGG (~41.5 MB).
```
grep -c HEROES2\.AGG /tmp/hero2.log
2896
```
I don't know how big chunks it reads this data.

A final note, dosbox-staging with built-in fluidsynth prevents from using external fluidsynth.

**Comment by dreamer** [comment]
@MasterO2 Yes, after 0.76.0; and I would *really, really* prefer to introduce Rust before we'll start introducing multithreading in all places we need it.

@kcgen Yes, I agree that making mixer multi-threaded (perhaps even with each channel operating in it's own thread) is a long-term goal, but we don't need it to have somewhat-usable FluidSynth support. Right now mixer callback is blocked for the whole time of FS synthesis; after we'll move it to separate thread it will be blocked only waiting for FS to finish the job and consume the buffer (synthesis will be able to start as soon as sysex commands will start arriving.

So we'll have usable FS MIDI, probably usable to the same level as emulating midi via GUS ULTRAMID is right now (which also results in radically reducing our emulation speed, but rarely anyone complains about it).

**Comment by dreamer** [comment]
> A final note, dosbox-staging with built-in fluidsynth prevents from using external fluidsynth.

What? How?

**Comment by kcgen** [comment]
@grapeli,

We certainly don't want you having to copy your games to tmpfs just to play them.

Fortunately Linux gives us some  knobs to achieve a "full preload on demand" behavior. 

1. Convert CD images (bin/cue) to directory of files.
    The allows Linux's filesystem's read-ahead setting to operate properly on a file-basis. Where as before, it will be incorrectly operating on the `.bin` file, because it doesn't have visibility into the ISO9660 files within.
1. Combine the CDROMs and game directory into one.
    This further collapses roughly ~240 MiB worth of duplicate files into a single file, and lets the read-ahead work more efficiently (ie: instead of reading the same file from CDROM and then local game directory, it just reads it once).
1. Defrag the resulting game directory using `e4defrag .`
1. Crank up the per-file read-ahead just for the block device storing your games.  Let's assume this is sda:
    `echo 49152 | sudo tee /sys/block/sda/queue/read_ahead_kb`.
    This ensures that the entire `HEROES2.AGG` will be moved into data-cache the very first byte that's read (which is right when you depart the castle).
1. Crank up the block-device read-ahead, which means that adjacent files during read will also be pulled into data cache (which are now likely to be other HOMM2 game files, given we've defragged the directory)
    ` sudo blockdev --setra 49152 /dev/sda`

The good news about both the file-system and block read-ahead is that they operate asynchronously beyond the original read request.  So when HOMM2 asks for a couple bytes from the AGG file, those bytes are returned immediately as soon as they're read - meanwhile the block and filesystem read-ahead carry on in parallel.  So it's the best of both worlds.

Let me know when you get this:
- http://162.156.79.96:8001/grapeli/h2rework.tar.zst
- just adjust the soundfont path inside its `dosbox.conf` (I didn't include a soundfont)
- Once received, I will remove this file.

**Comment by grapeli** [comment]
> > A final note, dosbox-staging with built-in fluidsynth prevents from using external fluidsynth.
> 
> @grapeli What? How? You need to configure sequencer port the same way as it always worked.

```
[midi]
mididevice = alsa
midiconfig = 128:0
```
It works. Oops. I set it to default.

**Comment by dreamer** [comment]
For anyone reading previous @kcgen's post: of course, this is only a temporary solution - we know users outside of Linux don't have sofisticated options of configuring readahead or tweaking filesystems.

Our goal is to have built-in FS usable on Windows and macOS as well. BTW, users who want to play HoMM2 - this game provides CD audio, so MIDI is not absolutely necessary; but it is very good stress-test.

**Comment by kcgen** [comment]
@dreamer, we might not be able to solve high latency storage problems though; unless we include our own embedded read-ahead mechanism.

**Comment by kcgen** [comment]
> @kcgen Today I don't have time for further analysis.

No worries. 

> I ran all tests on XFS.

Great.
 - The filesystem read-ahead and block read-ahead are agnostic of the specific filesystem, and will work just as well with XFS.  
 - For defrag, use: `find . -type f | sudo xargs xfs_fsr -v`

**Comment by realnc** [comment]
When you run fluidsynth in its own thread, you could have it output audio outside the dosbox mixer. Basically what the original fsynth dosbox patch did, except you'd keep its volume controllable by the dosbox mixer. This way, fsynth would behave similar to running it as a stand-alone client: It constantly renders audio and outputs to its own audio device handle, without blocking anything, and plays the MIDI events sent to it by dosbox.

**Comment by grapeli** [comment]
@kcgen 
Today I defragmented the entire xfs filesystem. I will try to make even better tuning.
I quickly checked under ext4. Not much better or the same (more repetitions required to be sure).
I downloaded the test package.

**Comment by kcgen** [comment]
Roger than @grapeli. Hopefully the two read-ahead settings make a notable difference.

**Comment by kcgen** [comment]
> When you run fluidsynth in its own thread, you could have it output audio outside the dosbox mixer.

@realnc , my understanding is that sdl only holds a single instance for a given audio device. So, even though the mixer and (hypothetical) theaded-FS could each hold their own pointer to the same audio device, under the hood sdl is only playing audio for one global buffer.

In other words, you can't write two separate buffers of audio (threaded and overlapping in time), and expect sdl to mix them into a single stream. 

This limitation drove the original need for Dosbox's mixer; otherwise each audio channel could have independently written its buffer straight into sdl and let it do the mixing.

(I know this first hand.. I tried to do it in the early version of w/ the CDDA patch 😅, which in turn drove me to using sdl_mixer-X, before giving up and using the dosbox mixer). 

OpenAL on the other hand does allow many simultaneous inbound streams into the same audio device, and it performs the mixing for you. So this FS push might revive our exploration into using OpenAL (or something like it).

**Comment by grapeli** [comment]
> 4. Crank up the per-file read-ahead just for the block device storing your games.  Let's assume this is sda:
>    `echo 49152 | sudo tee /sys/block/sda/queue/read_ahead_kb`.

> 5. Crank up the block-device read-ahead, which means that adjacent files during read will also be pulled into data cache (which are now likely to be other HOMM2 game files, given we've defragged the directory)
>    ` sudo blockdev --setra 49152 /dev/sda`

These two suggestions are quite good. It reduces the number of holes in the sound stream by half (first case), but more importantly they are located only in the first phase in the initial 20-25 seconds, the rest of the test is OK (before that, even in the final phase there were losses).

**Comment by kcgen** [comment]
@grapeli , that's good news; those settings should help slow seeking HDD's by "front-loading the latency-pain"; but like you noticed, it's still not perfect.  Let's try harder.

HOMM2's MIDI data is contained inside the `data/heroes2.agg` file, so we fine-tune this work-around a bit more.

Save the following as `start.sh` with execute permissions along side the test package's `dosbox.conf`.
Curious if that makes a difference:

``` shell
#!/bin/bash
set -xeu

# Ensure we're running inside the script directory
cd "$(dirname "$0")"

# Find and preload the data files containing latency-sensitive
# MIDI data. The pattern gets both the HOMM2 original campaign
# and expansion pack.
find heroes2/data -iname 'hero*.agg' \
 | xargs -i dd if="{}" of=/dev/null bs=1M

# Launch dosbox
/path/to/dosbox-binary -conf -userconf dosbox.conf 
```

If that works, I can envision a very easy way to move this functionality into dosbox, and a list of latency-sensitive files requiring preload provided via the conf file. It wouldn't actually hold them in memory (which is wasteful) and instead leverage the data-cache present in all modern operating system.

**Comment by realnc** [comment]
> my understanding is that sdl only holds a single instance for a given audio device. So, even though the mixer and (hypothetical) theaded-FS could each hold their own pointer to the same audio device, under the hood sdl is only playing audio for one global buffer.

Fluidsynth outputs audio directly through ALSA or PulseAudio (or whatever the equivalent is on other OSes.) SDL is an optional audio driver in fsync, but it's not the default.

This is how the original fsynth patch does it. This results in two audio streams showing up in the OS mixer when running dosbox. The OS in this case is doing the mixing, just as if two different processes are using the same audio device. Also, it looks kinda weird when you get two OS mixer sliders for the same app. It's not the best solution, but it's easy to do.

Another somewhat easy solution is to have fsynth only render audio in the audio thread. That would be `MIXER_CallBack()` in `hardware/mixer.cpp`, which is called by SDL in a different thread. This could potentially also improve rendering perf as well, because that way fsynth will render larger chunks each time, maybe getting better CPU cache utilization. The audio thread isn't doing anything else other than mixing right now, so it should have enough CPU time to both render and mix.

**Comment by kcgen** [comment]
Thanks @realnc. Yes, we wanted FluidSynth to use the same audio architecture as the other emulated audio devices in dosbox; but this might require an exception. 

Your second option would get my vote, and I certainly agree that larger audio rendering chunks would be much more efficient.

**Comment by dreamer** [comment]
@realnc 

> When you run fluidsynth in its own thread, you could have it output audio outside the dosbox mixer. Basically what the original fsynth dosbox patch did, except you'd keep its volume controllable by the dosbox mixer. This way, fsynth would behave similar to running it as a stand-alone client: It constantly renders audio and outputs to its own audio device handle, without blocking anything, and plays the MIDI events sent to it by dosbox.

We know, that's what we want to accomplish.

- Current solution is bad, because it synthesizes sound in the main thread
- Solution used by original patch is bad because it makes it impossible to control / postprocess FS output the way we can do right now

Ergo, we need to put synthesis into it's own thread (and let FS potentially split it into more threads internally, if user wants to). FS maintainer explicitly advised sythesis thread should use higher priority and not be responsible for other tasks (and I agree - some soundfonts use ridiculously big samples, and will need a lot of horsepower to process). 

> Another somewhat easy solution is to have fsynth only render audio in the audio thread. That would be MIXER_CallBack() in hardware/mixer.cpp, which is called by SDL in a different thread. (…)

Treating a single channel differently this way would be somewhat hacky solution in my opinion… But yeah, this is an option as well. I think putting FS into it's own thread will give a bit better results, but the proof is in the pudding as they say :) 

There's one more solution - I don't want to use it, because it has multiple other drawbacks, but listing it just for the sake of having complete picture. We could replace DOSBox mixer with SDL2_mixer library - it has built-in support for MIDI streams via Timidity… But it seems like SDL2_mixer has some issues, and I really prefer FluidSynth 2.x over Timidity (because it's maintained and AFAIK it's the only synth supporting the full set of SF2 features).

**Comment by realnc** [comment]
> We could replace DOSBox mixer with SDL2_mixer

That won't work. SDL_mixer can play MIDI files but not MIDI events. Also, SDL_mixer isn't actually intended for mixing audio. It's for decoding and playing various audio files (like WAV, Vorbis, etc.) The "mixer" in its name is somewhat unfortunate.

**Comment by grapeli** [comment]
> Then launch the game with it.
> 
> ```shell
> #!/bin/bash
> set -xeu
> 
> # Ensure we're running inside the script directory
> cd "$(dirname "$0")"
> 
> # Find and preload the data files containing latency-sensitive
> # MIDI data. The pattern gets both the HOMM2 original campaign
> # and expansion pack.
> find heroes2/data -iname 'hero*.agg' \
>  | xargs -i dd if="{}" of=/dev/null bs=1M
> 
> # Launch dosbox
> /path/to/dosbox-binary -conf -userconf dosbox.conf 
> ```
> 

It reduces to about 4-6.
There are also two files in the startup sequence that are open for writing and are modified. Autosave.gm1 file, after opening and each time you click the hourglass (next day).

The 8GB flash drive (ext4) also fails this test (probably due to slow writing).

Regarding the configuration you proposed, I had to slightly modify it. The laptop, in addition to a weak HDD and CPU, also has poor integrated graphics (Gen5) compatible only with OpenGL 2.1.
`output = texture` prevents scaling.
```
[sdl]
output = opengl
windowresolution = 1366x768
[cpu]
cycles = 60000
```
Here's a recording.
[homm2.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5364877/homm2.flac.txt)
Likewise, most problems are at the beginning (when the I / O is loaded), then it's clean.

**Comment by realnc** [comment]
> Treating a single channel differently this way would be somewhat hacky solution in my opinion… But yeah, this is an option as well. I think putting FS into it's own thread will give a bit better results, but the proof is in the pudding as they say :)

I completely forgot: fsynth wouldn't be the exception. Munt should be treated the same way. Or any other MIDI synth, like BASSMIDI (if you decide to add that at some point.)

This makes sense even from a philosophical point of view. MIDI synths were always "asynchronous" devices back in the day. If you had an MT-32, a SoundCanvas, a Wave Blaster, or whatever else, the game would just send MIDI events over the MPU-401 port. The sound was rendered and played by those devices on their own. They would not stutter, hiccup, underrun, or anything like that. Replicating that 100% async behavior makes sense.

**Comment by kcgen** [comment]
@grapeli,

Maybe there are other files (besides the big `heroes*.agg`) that load a bit of functional code prior to playing the MIDI sequences.
 - Can you try symlinking just the two `.agg` files over to tmpfs?
 - If you still get stutters then we know there are some other files that should be preloaded.
 - You can then incrementally move more files over tmpfs, each time flushing and retesting, and hopefully narrow down to exact files that need to be preloaded to prevent the stuttering.

Other thoughts:
If your system is starved for memory (or if many applications are idle and soaking up memory or swap), then there will be a lot of roll-over eviction where earlier read data is quickly evicted to make way for newer reads (even if we're only talking about 250MB of game data).  Memory starved systems also can behave very poorly when it comes to swapping out active application memory to disk; paging it back in is very slow and always visible to the user.  There are lots of knobs to tune this behavior, but at this point it's no longer related to FluidSynth - so perhaps we could chat over Discord via PM.

Regarding slow CPU and GPU -- given the game plays flawless when is moved to tmpfs, I think your CPU and GPU are fine.  dosbox doesn't need much horse-power to hit 30000 cycles (my $30 raspberry pi plays HOMM2 flawlessly w/ MIDI at 30k cycles); so I think the problem is isolated to your system's memory and IO subsystems.

**Comment by grapeli** [comment]
@kcgen 
I played HoMM2 a little (maybe too long). If I play it only with the music and sound turned off.
40,000 cycles  (definitely) are not sufficiently smooth on my laptop in my opinion.
Broken Alliance I pass in 31 days at impossible (it's a very decent result).

I will still be testing.

**Comment by grapeli** [comment]
The configuration is the same.
```
[sblaster]
sbtype  = none
```
Cache cleaned every time.
Switch to fullscreen 3-4 times. Next.
```
dir
cd heroes2
dir > list
del list
exit
```
[ece.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5373568/ece.flac.txt)
[staging.flac.txt](https://github.com/dosbox-staging/dosbox-staging/files/5373569/staging.flac.txt)

To make the difference visible, I increased the amplitude of the sound (-90dB).
How to explain this lack of fluidity in the sound source in dosbox-staging?
What is going on.

**Comment by kcgen** [comment]
Hi @grapeli, 

ECE applies the FluidSynth patch as originally authored by @realnc, who explained that the patch makes FluidSynth write its generated audio out to the host OS using FluidSynth's own audio drivers for alsa / PulseAudio / CoreAudio (macOS)  / Windows Audio. This audio output flow can happen asynchronously and in parallel to whatever dosbox is doing internally.

This differs from dosbox-staging's integration of @realnc's patch, which modified FluidSynth's audio output pathway to follow the same path as all the other emulated audio devices in dosbox, which is: raw PCM samples -> dosbox's mixer -> SDL -> host audio interface. All of this is driven in a 1ms synchronous and serialized loop.

Due to the lack of threading and output asynchronousity in the latter implementation, timing becomes absolutely critical.  That is, all the PCM samples (from all the emulated audio sources) need to be computed on the spot in time to fill 1ms worth of buffer, then mixed, and then written to SDL.. all within 1ms. If there's any lag (including lag caused by other parts of the emulator not related to audio), there will be an audible gap in the audio.

**Comment by dreamer** [comment]
@kcgen I thought ECE applies an older version of the patch, still using (now being deprecated) FluidSynth 1.x?

@grapeli Regarding `IO_ReadB` - do you build dosbox-staging yourself? If yes, then remember to use `-DNDEBUG` and `-O3` (as README.md suggests). Last time I checked we had no performance regressions compared to ECE, but well, this kind of regressions is hard to catch.

**Comment by kcgen** [comment]
> @kcgen I thought ECE applies an older version of the patch, still using (now being deprecated) FluidSynth 1.x?

Good point -- maybe ECE's patch is prior to @realnc's updates?
It looks like it includes work from @Kisai (Vogons.org forum). 

I'm _guessing_ it still using 1.x looking at `configure.ac`:

``` C
configure.ac:AC_CHECK_HEADER(fluidsynth.h, have_fluidsynth_h=yes, have_fluidsynth_h=no, )
configure.ac:AC_MSG_CHECKING(for FluidSynth support)
configure.ac:if test x$have_fluidsynth_lib = xyes -a x$have_fluidsynth_h = xyes ; then
configure.ac:  AC_DEFINE(C_FLUIDSYNTH,1)
configure.ac:  LIBS="$LIBS -lfluidsynth"
configure.ac:  AC_MSG_WARN([fluidsynth MIDI synthesis not available])
```

That said, the patch is definitely using (and requires) FluidSynth out audio output driver, as opposed to the channel -> mixer -> SDL pathway:

``` C++
                adriver = new_fluid_audio_driver(settings, synth);
                if (!adriver) {
                        LOG_MSG("MIDI:fluidsynth: Can't create audio driver");
                        delete_fluid_synth(synth);
                        delete_fluid_settings(settings);
                        return false;
                }
```
Regarding `IO_ReadB`: staging's IO handlers (approximately 270 of them) are stored in an STL  template ordered map, which offers ~O(1) lookup times. This change let us go from using an a hardcoded 6 MiB array (of mostly blank values) to a mere ~8 KiB index.

Speed-wise, there are effectively the same (the 6 MiB doesn't fit in most L2 caches so can take a hit paging it in while the ordered map might have to lookup up two values instead of one - but has very high odds of fitting entirely in L1 cache),  but in either case, these lookup times are dwarfed by the deliberate delay inserted into the IO handler code to prevent it going too fast (and likely getting divide-by-zeros and such):

![2020-10-13_17-10](https://user-images.githubusercontent.com/1557255/95928726-0325a480-0d77-11eb-89cd-f379fa882aeb.png)

But there's another thing at play here: the ordered map is an STL template customized at compile time. When you build with profile instrumentation, then the order map code itself becomes instrumented (which is great!) but at the same time, lookups become much slower due to that instrumentation.  Where as the 6 MiB raw array isn't going to have that instrumentation. So that's going to act as "balast weight" in the benchmark that disappears in a non-instrumented build.

**Comment by grapeli** [comment]
> > @kcgen I thought ECE applies an older version of the patch, still using (now being deprecated) FluidSynth 1.x?
> 
> Good point -- maybe ECE's patch is prior to @realnc's updates?
> It looks like it includes work from @kisai (Vogons.org forum).
> 
> That said, the patch is definitely using (and requires) FluidSynth's own audio output driver as opposed to the channel -> mixer -> SDL pathway (as there is no channel callback or add_samples() call):

@kcgen 
Exactly. The default midi config (ECE) was unchanged (`fluid.gain=.6`). Applying changes to the mixer  has no effect (`mixer fsynth 500`, hence the signal level is very low).

@dreamer 
Regarding IO_ReadB. There is no need to pay much attention to this. Perf top was made by hand, so it may have an error.

**Comment by grapeli** [comment]
I checked **dosbox-staging** under Windows 10 (same hardware). Configuration similar, soundfont used the same (FluidR3_GM.sf2), cold cache, HoMM2.
dosbox-staging-windows-x64-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)
dosbox-staging-windows-x86-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)

Then the x86-32 version under wine-5.19 under linux.
dosbox-staging-windows-x86-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)

The Linux version of dosbox-staging doesn't work (for me) properly with the built-in fluidsynth. I can rule out Alsa, I put it under Jack and it's the same.

**Comment by dreamer** [comment]
> v0.76.0-alpha-871-g51084

WTF, this is a PR build, not a build from our master branch… We explicitly filter development builds to include only those from master branch… WTF, GitHub REST API?

(as for FluidSynth, functionally it's the same as 93c7d6fe0d33e4a4bdbfcc0490a94d7fc027844f / current master branch)

> The Linux version of dosbox-staging doesn't work (for me) properly with the built-in fluidsynth. I can rule out Alsa, I put it under Jack and it's the same.

We don't use Jack API, and SDL does not support it either. Can you add more details about what doesn't work? Do you test on a distro that packaged FluidSynth 2.x already? I tested builds from the master branch on Fedora 33 Beta and they worked fine (aside of sound hiccups, that should be addressed by #640, which I will test next).

**Comment by grapeli** [comment]
@dreamer 
SDL2 uses Jack (as opposed to SDL1). Of course.
SDL_AUDIODRIVER=jack ./dosbox.

I'm testing a master with fluidsynth built from [contrib](https://github.com/dosbox-staging/dosbox-staging/blob/master/contrib/static-fluidsynth/Makefile). I wrote what sound font I use, I will write again - FluidR3_GM.sf2. I have sent over 5 samples showing what kind of problems it has with the sound output. I can send another one. It is no different. I am using Archlinux.

**Comment by dreamer** [comment]
Jack is not listed as a supported audio driver on [SDL Wiki](https://wiki.libsdl.org/FAQUsingSDL), but the list says "it's partial", but I don't know if it's partial in regards to platforms or available backends.

@grapeli sorry, I haven't noticed you listed the soundfont and I thought you are reporting some new issue.

**Comment by grapeli** [comment]
@dreamer 
sdl2 2.0.6
https://hg.libsdl.org/SDL/rev/eea7f98a37e4
I don't need Fluidsyth from the repository. It contains a number of redundant dependencies - libinstpatch, libsndfile, readline, dbus, systemd, ladspa, pulse, etc. Dosbox-staging doesn't need them.

**Comment by kcgen** [comment]
> With fluidsynth-2.1.5-2 from the repository, the sound is also interrupted.

![2020-10-23_09-14](https://user-images.githubusercontent.com/1557255/97027896-3ce17280-1510-11eb-8f43-74914fc6ef27.png)

What tool are you using tho create this?  This is extremely insightful; thanks for sharing this @grapeli!

Yet the build from `contrib` provides _flawless sound (no interruptions)_ -- I wonder what the underlying reason it?

Even though our contrib is as lean and optimized as we can make it, the repo-provided library should (mostly) have dormant pathways to those unnecessary services. Perhaps it not optimized or includes debug instruments and checks?

Very interesting none the less.

**Comment by grapeli** [comment]
Very lame (no doubt it can be better).
`SDL_AUDIODRIVER="alsa" AUDIODEV="hw:Loopback,0" ./dosbox`

https://ffmpeg.org/ffmpeg-filters.html#showspectrum-1
`ffmpeg -f alsa -c:a pcm_s16le -ac 2 -ar 44100 -i hw:2,1,0 -f x11grab -framerate 30000/1001 -video_size 1366x768 -i :0.0 -filter_complex "[0:a]showspectrum=s=200x768:orientation=horizontal,drawbox=w=iw:h=ih[vol]; [1:v]scale=1366:-1[scaled]; [scaled][vol]overlay[v]" -map "[v]" -map 0:a -c:v libx264 -pix_fmt yuv420p -qp 0 -preset ultrafast -g 60 /tmp/capture.01.mkv`

https://ffmpeg.org/ffmpeg-filters.html#showvolume
`ffmpeg -f alsa -c:a pcm_s16le -ac 2 -ar 44100 -i hw:2,1,0 -f x11grab -framerate 30000/1001 -video_size 1366x768 -i :0.0 -filter_complex "[0:a]showvolume=w=1366:p=0.6,drawbox=w=iw:h=ih[vol]; [1:v]scale=1366:-1[scaled]; [scaled][vol]overlay[v]" -map "[v]" -map 0:a -c:v libx264 -pix_fmt yuv420p -qp 0 -preset ultrafast -g 60 /tmp/capture.02.mkv`

**Comment by grapeli** [comment]
> I checked **dosbox-staging** under Windows 10 (same hardware). Configuration similar, soundfont used the same (FluidR3_GM.sf2), cold cache, HoMM2.
> dosbox-staging-windows-x64-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)
> dosbox-staging-windows-x86-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)
> 

I tried to verify it again. Once again it turns out that my conclusions are too hasty. Sound is choppy in HoMM2 similar to Linux. It turns out that something else is heard on the headphones (speakers) and something else is obtained in the file with the captured sound (Ctrl-F6). I was fooled.
Win10 runs sluggishly on this hardware with hdd. It takes nearly 8 minutes to start up and restart the system. Tests under this system are not for me. I have no strength to fight him.

> Then the x86-32 version under wine-5.19 under linux.
> dosbox-staging-windows-x86-v0.76.0-alpha-871-g51084 - flawless sound (no interruptions)

The repeated wine test also fails.
I'm making a lot of confusion.

**Comment by dreamer** [comment]
@grapeli thanks for retest; I suggest you hold your horses a little bit until #640 will be in - that one improves the situation for me (so far tested with HoMM2 on Linux, using my old laptop).

Another thing, that seemingly improves the situation (but it might be placebo, I did no measurements yet) was changing to `synth_threads = 2` (it introduces 1 new thread for FS synthesis, but I am not sure if it actually improves things, as we call fluidsynth in 1ms long chunks, which might be too small for multithreading to give us benefits at the moment).

Interestingly, I noticed that GOG release was toying with mixer prebuffer value for this game…

And yeah - your experience with sound issues happening outside of dosbox-staging (only when testing with headphones but not when recording) are similar to mine.

**Comment by kcgen** [comment]
> your experience with sound issues happening outside of dosbox-staging (only when testing with headphones but not when recording) are similar to mine.

This is so subtle and tricky; but fantastic you are both reproduced this too.  Wish we could get to the bottom of what's sitting between our output stream to SDL (the recording) and you headphones.

**Comment by grapeli** [comment]
Midi sound interruptions with built-in fluidsynth occur:
-only from hdd, or a slow flash drive, flash card
-**only with a cold cache(and only then)**
-most often when performing an interaction in the form of clicking on the touchpad, pressing the mouse button, pressing the keyboard
-when switching to fullscreen (and vice versa), switching to another application window (and back).

Never when the HoMM2 directory is mounted on tmpfs.
Never with external fluidsynth.
I also tested dosbox-ece and dosbox-x with fluidsynth built in. The sound is smooth without any distortion.

I tested dosbox-staging in various variants. With static SDL2, with `synth_threads=2`. With different audio outputs (dummy, alsa, jack).
With different linux kernels. In various configurations from `CONFIG_PREEMPT_NONE` to `CONFIG_PREEMPT`, from `CONFIG_HZ_100` to `CONFIG_HZ_1000`. From distribution (archlinux) to [BMQ](https://gitlab.com/alfredchen/linux-prjc/-/releases/v5.9-prjc-r1), from LTS, longterm to mainline. The configuration and kernel version have no effect on this.

I use statically built SDL2 a lot. For example, a niche WM - **notion**, I use needs a [patch](https://bugzilla-attachments.libsdl.org/attachment.cgi?id=2810) to make SDL2 work properly in full screen.

**Comment by dreamer** [comment]
With #734 finished, I think it's time to close this issue :)

FluidSynth 2.x integration is not perfect yet, but it already provides the best out-of-the-box experience compared to any MIDI solution I tested with any DOSBox fork so far :)

We picked good defaults, support soundfont lookup in standard directories, gain configuration that's easy to correlate to a specific font, have updated documentation to reflect the changes, and more.

To enable FluidSynth MIDI, user needs to set:

```ini
[midi]
mididevice = fluidsynth
mpu401 = intelligent

[fluidsynth]
soundfont = filename.sf2  # file will be looked in current working dir, then
                          # in 'soundfonts' dir in config dir, and a number
                          # of global directories that might store .sf2 files.

# alternatively:
soundfont = ~/my-soundfonts/foobar.sf2  # relative, absolute paths and tilde-expansion
                                        # is supported

# alternatively:
soundfont = quiet-font.sf2 150  # gain is adjusted to 150%
soundfont = loud-font.sf2 30    # gain is reduced to 30%
```
Empty `soundfonts` dir is created for the user automatically, so it's clear where to store fonts for DOS games. Or they can be stored with the game itself (assuming user changed CWD to game directory).

The volume of FluidSynth MIDI channel (`FSYNTH`) can be adjusted via DOS `mixer` command (the same way as every other built-in emulation channel).

DOSBox Staging will look for `default.sf2` file by default, as this is the default name used by FluidSynth - on Linux systems it's usually a symbolic link to the newest version of Fluid_R3.sf2 (if it's installed).

@kcgen, @grapeli, @realnc, @nemo93, @bluddy
Thank you everyone for testing, let's continue improving the feature in #735 :)
